crates/grit-pattern-matcher/src/pattern/state.rs

```diff
@@ -95,8 +95,9 @@ impl<'a, Q: QueryContext> FileRegistry<'a, Q> {
     }
 
     /// If only the paths are available, create a FileRegistry with empty owners
-    /// This is *unsafe* if you do not later insert the appropriate owners before get_file_owner is called
-    pub fn new_from_paths(file_paths: Vec<&'a PathBuf>) -> Self {
+    /// This is a logic error if you do not later insert the appropriate owners before get_file_owner is called
+    pub fn new_from_paths(file_paths: impl Iterator<Item = &'a Path>) -> Self {
+        let file_paths: Vec<&Path> = file_paths.collect();
```

Comment:
On second thought, I wonder how useful this really is. If we collect into a `Vec` anyway, we may as well let the caller do it, so it's more explicit and we don't hide the allocation.

I'm thinking #591 was actually more about the request to use `&Path` instead of `&PathBuf`, than it was about the change to use an iterator or slice. What do you think, @ematipico?

Reply:
The upstream issue is caused by the fact that we don't use `PathBuf`/`Path` anymore, so when creating a vector of references for the library (this crate), we can't create a Vector of references because they don't live long a enough. 

Passing a vector of references is actually a weird API, because it's a owned collection of references. It doesn't play very well.

So if the caller needs to pass the data, it might as well be a `Vec<PathBuf>` at this point. No lifetimes

Reply:
I meant it more from the consideration that by accepting `&Path` we can already use it more easily with `camino` through the use of `as_std_path()`. The paths can remain references to avoid unnecessary cloning.

You’re right `Vec<PathBuf>` would be easier from a lifetime perspective, but it seems unnecessary from our perspective and it would introduce unnecessary cloning, so I don’t think we need to make that change.

But that leaves the question, do we pass the paths as a `Vec`, a slice, or an iterator. I changed it to an iterator now, but it seems unnecessary: It will be stored in the data structure as a `Vec` anyway, and it’s more idiomatic in Rust to let the caller do the allocation, which gives it the possibility to elide the allocation if an existing variable can be moved instead.

So I don’t actually agree that passing a vector of references is a weird API. It may be less common, but it can still be the right thing to do.